python/lsst/meas/algorithms/dynamicDetection.py

```diff
@@ -0,0 +1,210 @@
+        threshold.
+
+        Parameters
+        ----------
+        exposure : `lsst.afw.image.Exposure`
+            Exposure to process; DETECTED{,_NEGATIVE} mask plane will be
+            set in-place.
+        doSmooth : `bool`
+            If True, smooth the image before detection using a Gaussian
+            of width ``sigma``.
+        sigma : `float`, optional
+            Gaussian Sigma of PSF (pixels); used for smoothing and to grow
+            detections; if `None` then measure the sigma of the PSF of the
+            ``exposure``.
+        clearMask : `bool`
+            Clear both DETECTED and DETECTED_NEGATIVE planes before running
+            detection.
+
+        Return Struct contents
+        ----------------------
+        positive : `lsst.afw.detection.FootprintSet`
+            Positive polarity footprints (may be `None`)
+        negative : `lsst.afw.detection.FootprintSet`
+            Negative polarity footprints (may be `None`)
+        numPos : `int`
+            Number of footprints in positive or 0 if detection polarity was
+            negative.
+        numNeg : `int`
+            Number of footprints in negative or 0 if detection polarity was
+            positive.
+        background : `lsst.afw.math.BackgroundMI`
+            Re-estimated background.  `None` if
+            ``reEstimateBackground==False``.
+        factor : `float`
+            Multiplication factor applied to the configured detection
+            threshold.
+        """
+        maskedImage = exposure.maskedImage
+
+        # We need a seed for the RNG in the sky object generator, but we don't have
+        # access to the butler, which is what we'd normally use to get a seed for a
+        # particular piece of data. We need something that changes with the input
+        # (so we don't use the same sky objects for every input), but deterministic.
+        # It also needs to fit in a C++ 'int' so pybind doesn't choke on it.
+        seed = int(maskedImage.image.array.sum()) % (2**31 - 1)
+
+        if clearMask:
+            self.clearMask(maskedImage.mask)
+        else:
+            oldMask = maskedImage.mask.array & maskedImage.mask.getPlaneBitMask(["DETECTED",
+                                                                                 "DETECTED_NEGATIVE"])
+
+        # Could potentially smooth with a wider kernel than the PSF in order to better pick up the
+        # wings of stars and galaxies, but for now sticking with the PSF as that's more simple.
+        psf = self.getPsf(exposure, sigma=sigma)
+        convolveResults = self.convolveImage(maskedImage, psf, doSmooth=doSmooth)
+        middle = convolveResults.middle
+        sigma = convolveResults.sigma
+
+        prelim = self.applyThreshold(middle, maskedImage.getBBox(), self.config.prelimThresholdFactor)
+
+        if self.config.doTempLocalBackground:
+            self.applyTempLocalBackground(exposure, middle, prelim)
```

Comment:
Do we need to run this on the preliminary detections?  If it just updates the peaks, it shouldn't affect the new threshold calculation.  If we're relying on it for the side-effect of subtracting the background from middle, maybe we should refactor a bit more so we can avoid wasting time updating the peaks here.